### PR TITLE
Max download size and timeout control for native download routines

### PR DESCRIPTION
**Pull request by pisajew**
Note that this requires 2 additional enviromnent parameters:
DOWNLOAD_TIMEOUT_SECONDS and MAXIMUM_DOWNLOAD_SIZE_KB

setting those parameters to 0 effectively disables this functionality but their absence will cause compilation errors for Android.

Resolves https://github.com/cds-snc/covid-shield-mobile/issues/527

### Testing proposal

- Setup a local server which handles /get-configuration and retrieve endpoints, with configurable delay and response size
- configure MAXIMUM_DOWNLOAD_SIZE_KB and DOWNLOAD_TIMEOUT_SECONDS so that both are exceeded by the server.
- configure the retrieval endpoint to point to your configured server.

**Example**

I'm using the following script:
`
#!/usr/bin/perl

use strict;

{
  package CovidWebServer;

  use HTTP::Server::Simple::CGI;
  use base qw(HTTP::Server::Simple::CGI);

  my $HUGE_FILE_SIZE = 10 * 1024 * 1024;
  my $SLOW_FILE_DELAY = 125;

  sub handle_request {
    my $self = shift;
    my $cgi = shift;
    my $path = $cgi->path_info();

    print "HTTP/1.0 200 OK\r\n";
    
    if($path =~ m|/backend/.?retrieve|) {
      print "Content-type: application/zip\r\n";
      print "\r\n";

      for(my $i = 0; $i < $HUGE_FILE_SIZE; $i++) {
        print  ' ';
      }
    } elsif($path =~ m|/backend/.?exposure-configuration|) {
      print "Content-type: application/json\r\n";
      print "\r\n";
      
      sleep($SLOW_FILE_DELAY);
      
      print q({
  "minimumRiskScore": 1,
  "attenuationLevelValues": [
    0,
    0,
    2,
    2,
    2,
    2,
    2,
    2
  ],
  "attenuationWeight": 50,
  "daysSinceLastExposureLevelValues": [
    0,
    1,
    1,
    1,
    1,
    1,
    1,
    1
  ],
  "daysSinceLastExposureWeight": 50,
  "durationLevelValues": [
    0,
    0,
    0,
    0,
    5,
    5,
    5,
    5
  ],
  "durationWeight": 50,
  "transmissionRiskLevelValues": [
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1
  ],
  "transmissionRiskWeight": 50
});
      
    }
  }
}

CovidWebServer->new(8080)->run();
`

and configure SUBMIT/RETRIEVE_URL to http://my_server_ip:8080/backend/

Launching app in debug mode it will be possible to see on console that requests are made and then fail after the limits are exceeded.

### CODE REVIEW

**Comment by timarney** [comment]
@obrien-j outside of the app team who should review this?

**Comment by timarney** [comment]
Received confirmation that the submitter of the flagged issue is happy with this.

**Comment by timarney** [comment]
@pisajew can you add some notes with the PR on suggestions on how to test this?

File: .env.example [hunk 1]
```diff
@@ -10,3 +10,6 @@ MCC_CODE=302
 
 TEST_MODE=false
 MOCK_SERVER=false
+DOWNLOAD_TIMEOUT_SECONDS=120
+MAXIMUM_DOWNLOAD_SIZE_KB=10000
```
Comment (henrytao-me) [inline .env.example:14]: This is roughly 10MB. I download last 14 days keys from server. It's 13KB for 635 keys. So, 10MB will be roughly 400k keys which is more than enough, I think. 👍

File: src/services/BackendService/BackendService.ts [hunk 2]
```diff
@@ -56,7 +58,19 @@ export class BackendService implements BackendInterface {
     const region = 'CA';
     const exposureConfigurationUrl = `${this.retrieveUrl}/exposure-configuration/${region}.json`;
     captureMessage('getExposureConfiguration', {exposureConfigurationUrl});
-    return (await fetch(exposureConfigurationUrl, FETCH_HEADERS)).json();
+    const timeoutSecs = Number(Config.DOWNLOAD_TIMEOUT_SECONDS);
+
+    const abortController = new AbortController();
+    const opts = Object.assign(FETCH_HEADERS, {signal: abortController.signal});
+
+    if (timeoutSecs > 0) {
```
Comment (henrytao-me) [inline src/services/BackendService/BackendService.ts:66]: `getExposureConfiguration` is the only `fetch` that we use in js layer. We can move it to native layer via CovidShield Module for consistency.

File: android/app/src/main/java/app/covidshield/module/CovidShieldModule.kt [hunk 3]
```diff
@@ -15,11 +16,18 @@ import java.io.File
 import java.io.IOException
 import java.security.SecureRandom
 import java.util.*
+import java.util.concurrent.TimeUnit
 import kotlin.coroutines.CoroutineContext
 
 class CovidShieldModule(context: ReactApplicationContext) : ReactContextBaseJavaModule(context), CoroutineScope {
 
-    private val okHttpClient by lazy { OkHttpClient() }
+
```
Comment (henrytao-me) [inline android/app/src/main/java/app/covidshield/module/CovidShieldModule.kt:24]: 👾

File: android/app/src/main/java/app/covidshield/module/DownloadInterceptor.kt [hunk 4]
```diff
@@ -0,0 +1,62 @@
+package app.covidshield.module
+
+import okhttp3.Interceptor
+import okhttp3.MediaType
+import okhttp3.Response
+import okhttp3.ResponseBody
+import okhttp3.Call
+import okio.*
+
+class DownloadInterceptor(val maxBytes: Long): Interceptor {
```
Comment (henrytao-me) [inline android/app/src/main/java/app/covidshield/module/DownloadInterceptor.kt:10]: `maxBytes` should be private.

File: src/services/BackendService/BackendService.ts [hunk 5]
```diff
@@ -56,7 +58,19 @@ export class BackendService implements BackendInterface {
...
+      clearTimeout(this.configFetchTimeout);
+      this.configFetchTimeout = setTimeout(() => {
```
Comment (henrytao-me) [inline src/services/BackendService/BackendService.ts:68]: setTimeout for too long in js layer is not recommend. We should move this to native layer then.

File: src/services/BackendService/BackendService.ts (hunk 2 above)
Comment (gavrix) [reply src/services/BackendService/BackendService.ts:66]: I would leave as is.

**Comment by timarney** [comment]
cc: @dsamojlenko 

> Note that this requires 2 additional enviromnent parameters:
DOWNLOAD_TIMEOUT_SECONDS and MAXIMUM_DOWNLOAD_SIZE_KB

File: src/services/BackendService/BackendService.ts (hunk 2 above)
Comment (pisajew) [reply src/services/BackendService/BackendService.ts]: moved to native layer now

File: android/app/src/main/java/app/covidshield/module/CovidShieldModule.kt [hunk 6]
```diff
@@ -50,6 +60,20 @@ class CovidShieldModule(context: ReactApplicationContext) : ReactContextBaseJava
         }
     }
 
+    @ReactMethod
+    fun fetchExposureConfiguration(url: String, promise: Promise) {
+        promise.launch(this) {
+            val request = Request.Builder().url(url).build()
+            okHttpClient.newCall(request).execute().use { response ->
```
Comment (jeberhardt) [inline android/app/src/main/java/app/covidshield/module/CovidShieldModule.kt:67]: When merging this code, be sure to incorporate the `cache-control` code:

`.cacheControl(CacheControl.Builder().noStore().build())`

File: android/app/src/main/java/app/covidshield/module/CovidShieldModule.kt [hunk 7]
```diff
@@ -50,6 +60,20 @@ class CovidShieldModule(context: ReactApplicationContext) : ReactContextBaseJava
         }
     }
 
+    @ReactMethod
+    fun fetchExposureConfiguration(url: String, promise: Promise) {
+        promise.launch(this) {
+            val request = Request.Builder().url(url).build()
```
Comment (jeberhardt) [inline android/app/src/main/java/app/covidshield/module/CovidShieldModule.kt:66]: When merging this code, be sure to incorporate the `cache-control` code:

`.cacheControl(CacheControl.Builder().noStore().build())`

File: ios/CovidShield/CovidShield.m [hunk 8]
```diff
@@ -25,28 +27,59 @@ @implementation CovidShield
...
+                                             error: nil], nil);
+        } else {
+          successHandler(location);
+          [[NSFileManager defaultManager] removeItemAtURL:location error:nil];
+        }
+      }] downloadTaskWithURL: url];
+}
+
+- (NSURLSession *)configuredSessionWithCompletionHandler: (void (^)(NSURL *, NSURLResponse *, NSError *)) completion {
+  NSURLSessionConfiguration *sconf = [NSURLSessionConfiguration defaultSessionConfiguration];
```
Comment (jeberhardt) [inline ios/CovidShield/CovidShield.m:72]: Please change this to use:
`[NSURLSessionConfiguration ephemeralSessionConfiguration]`

This was done in a previous PR, to ensure that downloaded items won't be stored in cache. I'm just pointing this out so it doesn't get lost in the merge.

https://github.com/cds-snc/covid-alert-app/pull/859

**Comment by jeberhardt** [comment]
When merging this PR, please be sure that the changes made in these PRs are not undone. They revolve around setting `cacheControl` for both Android and iOS at the native level. See my comments in the code.

https://github.com/cds-snc/covid-alert-app/pull/859
https://github.com/cds-snc/covid-alert-app/pull/858

File: ios/CovidShield/CovidShield.m (hunk 8 above)
Comment (pisajew) [reply ios/CovidShield/CovidShield.m]: addressed by f5d49e3c5f50abb94abd7923cbfd084a4355c270

**Comment by timarney** [comment]
Closes this pending further discussion we'll re-open as needed.

**Pull request closed without merging**